cpp/arcticdb/async/async_store.hpp

```diff
@@ -280,36 +280,35 @@ class AsyncStore : public Store {
         return res;
     }
 
-    std::vector<VariantKey> batch_read_compressed(
+    folly::Future<std::vector<VariantKey>> batch_read_compressed(
         std::vector<entity::VariantKey> &&keys,
         std::vector<ReadContinuation> &&continuations,
         const BatchReadArgs & args) override {
         util::check(keys.size() == continuations.size(),
                     "keys and continuations must has then same number of elements");
-        std::vector<folly::Future<VariantKey>> batch;
-        batch.reserve(args.batch_size_);
-        std::vector<VariantKey> res;
-        res.reserve(keys.size());
-        for (std::size_t i = 0; i < keys.size(); ++i) {
-            auto &key = keys[i];
-            auto &cont = continuations[i];
-            batch.push_back(
-                async::submit_io_task(ReadCompressedTask(key, library_, storage::ReadKeyOpts{}))
-                    .via(&async::cpu_executor())
-                    .thenValue(SegmentFunctionTask{std::move(cont)}));
-
-            if(batch.size() == args.batch_size_) {
-                auto vec = folly::collect(batch).get();
-                res.insert(end(res), std::make_move_iterator(std::begin(vec)), std::make_move_iterator(std::end(vec)));
-                batch.clear();
-            }
-        }
-
-        if(!batch.empty()) {
-            auto vec = folly::collect(batch).get();
-            res.insert(end(res), std::make_move_iterator(std::begin(vec)), std::make_move_iterator(std::end(vec)));
+        
+        
+        std::vector<std::tuple<entity::VariantKey, int>> keys_and_continuations_idx(keys.size());
+        std::transform(std::make_move_iterator(keys.cbegin()),
+                    std::make_move_iterator(keys.cend()),
+                    keys_and_continuations_idx.begin(),
+                    [](auto&& key_val) {
+            return std::make_tuple(std::move(key_val),0);
+        });
+        
+        for (auto key_and_idx = keys_and_continuations_idx.begin(); key_and_idx < keys_and_continuations_idx.end(); ++key_and_idx) {
+            std::get<1>(*key_and_idx) = key_and_idx - keys_and_continuations_idx.cbegin();
```

Comment:
So the 1th element of the tuple is just the position/ordinal?

Try `folly/container/Enumerate.h`, or better yet, directly zip `continuations` in using something like [this](https://stackoverflow.com/a/51032349/499922) and avoid the index altogether.

Reply:
that's exactly what I tried in the first place, but std::pair doesn't cope well with folly::functions